### ISSUE TITLE
Undefined symbol when building from source

### ISSUE THREAD

**Issue by brokencuph** [taichi-dev/taichi#7820]
**Describe the bug**
When I build taichi from source following [this documentation](https://docs.taichi-lang.org/docs/dev_install), the compilation succeeds without errors, but importing taichi brings undefined symbol error.

**To Reproduce**
Just importing taichi in any way (e.g. running an example) after building from source
```py
import taichi
```

**Log/Screenshots**
It shows:
```
>>> import taichi
Share object taichi_python import failed, check this page for possible solutions:
https://docs.taichi-lang.org/docs/install
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/hyj/taichi/python/taichi/__init__.py", line 3, in <module>
    from taichi._funcs import *
  File "/home/hyj/taichi/python/taichi/_funcs.py", line 3, in <module>
    from taichi.lang import impl, ops
  File "/home/hyj/taichi/python/taichi/lang/__init__.py", line 1, in <module>
    from taichi.lang import impl, simt
  File "/home/hyj/taichi/python/taichi/lang/impl.py", line 6, in <module>
    from taichi._lib import core as _ti_core
  File "/home/hyj/taichi/python/taichi/_lib/__init__.py", line 1, in <module>
    from taichi._lib.utils import ti_python_core as core
  File "/home/hyj/taichi/python/taichi/_lib/utils.py", line 110, in <module>
    ti_python_core = import_ti_python_core()
  File "/home/hyj/taichi/python/taichi/_lib/utils.py", line 55, in import_ti_python_core
    raise e from None
  File "/home/hyj/taichi/python/taichi/_lib/utils.py", line 45, in import_ti_python_core
    from taichi._lib.core import \
ImportError: /home/hyj/taichi/python/taichi/_lib/core/taichi_python.cpython-310-x86_64-linux-gnu.so: undefined symbol: _ZN6taichi4lang13window_system20glfw_context_releaseEv
```

**Additional comments**
OS: Ubuntu 22.04
Python version: 3.10
Clang version: 
```
Ubuntu clang version 14.0.0-1ubuntu1
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```
LLVM: use the downloaded provided prebuilt LLVM

Inspecting the object files (`taichi/rhi/libti_device_api.a`) shows that relevant function symbols being hidden:
```
$ readelf -Cs --wide libti_device_api.a | grep glfw
     2: 0000000000000000     0 SECTION LOCAL  DEFAULT    3 .text._ZN6taichi4lang13window_system20glfw_context_acquireEv
     3: 0000000000000000    48 OBJECT  LOCAL  DEFAULT   16 taichi::lang::window_system::glfw_state
     4: 0000000000000000   183 FUNC    LOCAL  DEFAULT    6 taichi::lang::window_system::glfw_error_callback(int, char const*)
     5: 0000000000000000     0 SECTION LOCAL  DEFAULT    5 .gcc_except_table._ZN6taichi4lang13window_system20glfw_context_acquireEv
     7: 0000000000000000     0 SECTION LOCAL  DEFAULT    6 .text._ZN6taichi4lang13window_systemL19glfw_error_callbackEiPKc
    10: 0000000000000000     0 SECTION LOCAL  DEFAULT    8 .text._ZN6taichi4lang13window_system20glfw_context_releaseEv
    13: 000000000000007e    57 OBJECT  LOCAL  DEFAULT   14 .L__PRETTY_FUNCTION__._ZN6taichi4lang13window_system20glfw_context_releaseEv
    14: 0000000000000000     0 SECTION LOCAL  DEFAULT   10 .gcc_except_table._ZN6taichi4lang13window_system20glfw_context_releaseEv
    29: 0000000000000000     0 SECTION LOCAL  DEFAULT   16 .bss._ZN6taichi4lang13window_systemL10glfw_stateE
    30: 0000000000000000   115 FUNC    GLOBAL HIDDEN     3 taichi::lang::window_system::glfw_context_acquire()
    32: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND glfwInit
    33: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND glfwSetErrorCallback
    45: 0000000000000000   116 FUNC    GLOBAL HIDDEN     8 taichi::lang::window_system::glfw_context_release()
    46: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND glfwTerminate
```
And in the resulting .so it is UND:
```
$ readelf --dyn-syms -C --wide taichi_python.cpython-310-x86_64-linux-gnu.so | grep glfw
    20: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND taichi::lang::window_system::glfw_context_release()
   227: 0000000000000000     0 NOTYPE  GLOBAL DEFAULT  UND taichi::lang::window_system::glfw_context_acquire()
```
If I manually force the symbols not to be hidden, it only takes effect on the .a file, but no change to the .so.

**Comment by turbo0628** [comment]
Also reproduced the same error on a freshly install 20.04

**Comment by jim19930609** [comment]
Linux `ld` requires strict ordering when performing symbol resolution, so if there's underlying circular dependencies then it's likely gonna cause this undefined symbol problem. Did some minor adjustments to RHI build structure but still fairly hacky due to the way `window_system` being organized.

**Comment by brokencuph** [comment]
Thanks for quick responding! It was the first time I tried to compile Taichi from source, and happen to find this problem.

**Comment by ailzhang** [comment]
Fixed by #7827 , closing ;)
